### PR TITLE
feat: add vm extension related components & enable transpiling ESM js file for jest

### PR DESCRIPTION
**Pull request by wanghoppe**
image

image

### CODE REVIEW

File: packages/react/src/vm-extension/mock-vme-ext-items.ts
Comment (gingi) [inline packages/react/src/vm-extension/mock-vme-ext-items.ts:1]: Should this be named `mock-vm-ext-items.ts`?

**Review by gingi — DISMISSED** [review]
Looks good. One small question about a file name; if by intention then ignore.

File: packages/react/src/vm-extension/mock-vme-ext-items.ts
Comment (wanghoppe) [reply packages/react/src/vm-extension/mock-vme-ext-items.ts:1]: Yeah, thanks for catching that, will fix.

File: packages/bonito-ui/src/components/property/property-field.tsx [hunk 1]
```diff
@@ -6,6 +6,8 @@ import { copyToClipboard } from "@azure/bonito-core";
 export interface PropertyFieldProps<T> {
     label?: string;
     value?: T;
+    hideCopyButton?: boolean;
+    labalStyle?: React.CSSProperties;
```
Comment (dpwatrous) [inline packages/bonito-ui/src/components/property/property-field.tsx:10]: labal -> label

File: packages/bonito-ui/src/environment/mock-browser-environment.ts [hunk 2]
```diff
@@ -38,6 +38,24 @@ export class MockBrowserEnvironment
...
+        // some JSDOM APIs are not implemented, so we need to mock them
+        // to make importing monaco-editor works
+        // https://jestjs.io/docs/manual-mocks#mocking-methods-which-are-not-implemented-in-jsdom
+        Object.defineProperty(window, "matchMedia", {
+            writable: true,
+            value: jest.fn().mockImplementation((query) => ({
+                matches: false,
+                media: query,
+                onchange: null,
+                addListener: jest.fn(), // deprecated
```
Comment (dpwatrous) [inline packages/bonito-ui/src/environment/mock-browser-environment.ts:51]: Let's move this into Jest-specific code. MockBrowserEnvironment can also be used for the playground, and doesn't require Jest. It also runs in a real browser, so doesn't need these stubs.

File: packages/playground/src/demo/display/vm-extension/vm-extension-list-demo.tsx [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    // const [panelOpen, setPanelOpen] = React.useState<boolean>(false);
+    // const [selectedExtension, setSelectedExtension] =
+    //     React.useState<VmExtItem>();
+
+    return (
+        <DemoPane title="VM Extension List">
+            <VmExtensionList
+                extensions={allExtItems}
+                loading={false}
+                // onItemClick={(item) => {
```
Comment (dpwatrous) [inline packages/playground/src/demo/display/vm-extension/vm-extension-list-demo.tsx:20]: Can this be removed?

File: packages/react/src/vm-extension/__tests__/node-vm-ext-list.spec.tsx [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+    it("should render vm extension lists", async () => {
+        const { getByText } = render(
+            <NodeVMExtList
+                accountEndpoint="dummy_endpoint"
+                poolId={"dummy_pool"}
+                nodeId="tvmps_id1"
+            />
+        );
+        getByText(translate("lib.react.vmExtension.name"));
```
Comment (dpwatrous) [inline packages/react/src/vm-extension/__tests__/node-vm-ext-list.spec.tsx:20]: Can we add an accessibility test in here somewhere? Also in the pool extension list test.

File: packages/react/src/vm-extension/node-vm-ext-list.tsx [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    const nodeService: NodeService = React.useMemo(() => {
+        return inject(BatchDependencyName.NodeService);
+    }, []);
+
+    React.useEffect(() => {
+        setLoading(true);
+        nodeService
+            .listBatchNodeExtensions(accountEndpoint, poolId, nodeId)
+            .then((resList) => {
+                const exts = [];
```
Comment (dpwatrous) [inline packages/react/src/vm-extension/node-vm-ext-list.tsx:29]: Need to check if the component is still mounted by the time this runs. To track this, you can return a callback which sets a local variable. See action-form.tsx for an example.

This is a pattern that's really hard to get right consistently. Probably a good case for a custom hook.

**Review by dpwatrous — CHANGES_REQUESTED** [review]

File: packages/react/src/vm-extension/node-vm-ext-list.tsx (hunk 5 above)
Comment (dpwatrous) [reply packages/react/src/vm-extension/node-vm-ext-list.tsx:29]: To expand a bit on the issue, imagine the following hook:

```
useEffect(() => {
    loadSomeData()
        .then((data) => {
            setSomeState(data);
        });
}, [dependency1, dependency2])
```

Now imagine the hook is run multiple times in quick succession due to component rerendering. Every `loadSomeData()` promise might still be unresolved, and there's no guaranteed order in which they'll resolve. The callbacks will race, and the final data passed to `setSomeState()` will vary based on the order in which the callbacks fire.

Adding a check to see if the component is still mounted before setting state will resolve this:

```
useEffect(() => {
    let isMounted = true;
    loadSomeData()
        .then((data) => {
            if (isMounted) {
                setSomeState(data);
            }
        });
    return () => { isMounted = false; };
}, [dependency1, dependency2])
```

See also: https://react.dev/learn/synchronizing-with-effects#fetching-data

File: packages/bonito-core/config/tsconfig.test.json [hunk 6]
```diff
@@ -0,0 +1,8 @@
+{
+  "extends": "./tsconfig.common.json",
+  "compilerOptions": {
+    "tsBuildInfoFile": "../build/cache/ts.build.buildinfo",
```
[the remark this answers is not in the document]
Comment (wanghoppe) [reply packages/bonito-core/config/tsconfig.test.json]: renamed

File: util/common-config/jest-common.js [hunk 7]
```diff
@@ -55,6 +59,12 @@ module.exports = {
                 ],
             ],
             globals: {
+                "ts-jest": {
+                    // js-with-ts preset require "allowJs": true in tsconfig.json
+                    // so making a "tsconfig.test.json" to override it
+                    tsconfig: "config/tsconfig.test.json",
+                    isolatedModules: true,
```
[the remark this answers is not in the document]
Comment (wanghoppe) [reply util/common-config/jest-common.js:66]: Tested it only works in `globals` with `js-with-ts` preset

File: util/common-config/jest-common.js [hunk 8]
```diff
@@ -105,8 +119,15 @@ module.exports = {
         baseConfig.moduleNameMapper["@batch/ui-service/lib/(.*)$"] =
             "@batch/ui-service/lib-cjs/$1";
 
+        // css files
+        baseConfig.moduleNameMapper["^.+\\.css$"] = path.join(
+            __dirname,
+            "../mock-style.js"
```
[the remark this answers is not in the document]
Comment (wanghoppe) [reply util/common-config/jest-common.js:125]: Added comments

File: packages/bonito-ui/src/components/panel/panel.tsx [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+export interface PanelProps {
+    headerText?: string;
+    isOpen?: boolean;
+    onDismiss?: () => void;
+    isBlocking?: boolean;
+    footerButtons?: ButtonProps[];
+}
+
+export const Panel: React.FC<PropsWithChildren<PanelProps>> = (props) => {
```
[the remark this answers is not in the document]
Comment (wanghoppe) [reply packages/bonito-ui/src/components/panel/panel.tsx:14]: Added comments

File: packages/bonito-ui/src/components/property/__tests__/property-group.spec.tsx [hunk 10]
```diff
@@ -56,4 +57,28 @@ describe("PropertyGroup component", () => {
         expect(heading.getAttribute("aria-expanded")).toBe("false");
         expect(section.style.display).toBe("none");
     });
+
+    test('Should not be able to expand/collapse if "enableCollapse" is false', async () => {
+        await render(
+            <PropertyGroup title="Car" enableCollapse={false}>
```
[the remark this answers is not in the document]
Comment (wanghoppe) [reply packages/bonito-ui/src/components/property/__tests__/property-group.spec.tsx:63]: Done

File: packages/bonito-ui/src/environment/mock-browser-environment.ts (hunk 2 above)
Comment (wanghoppe) [reply packages/bonito-ui/src/environment/mock-browser-environment.ts]: moved jsdom mocks to setup-tests

File: packages/playground/src/demo/display/vm-extension/vm-extension-list-demo.tsx (hunk 3 above)
Comment (wanghoppe) [reply packages/playground/src/demo/display/vm-extension/vm-extension-list-demo.tsx]: Resolved

File: packages/react/src/vm-extension/__tests__/node-vm-ext-list.spec.tsx (hunk 4 above)
Comment (wanghoppe) [reply packages/react/src/vm-extension/__tests__/node-vm-ext-list.spec.tsx:20]: added for node and pool VM extension list

File: packages/bonito-ui/src/components/data-grid/data-grid.tsx [hunk 11]
```diff
@@ -102,7 +102,14 @@ export interface DataGridColumn {
     /**
      * Custom renderer for cell content, instead of the default text rendering.
      */
-    onRender?: (item?: unknown, index?: number, column?: IColumn) => unknown;
+    onRender?: (
+        // can't use unknown here because it's a function's parameter, typescript
+        // check it contravariantly
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        item: any,
```
[the remark this answers is not in the document]
Comment (wanghoppe) [reply packages/bonito-ui/src/components/data-grid/data-grid.tsx]: add generic type for data-grid

File: packages/bonito-core/src/__tests__/setup-tests.ts [hunk 12]
```diff
@@ -1,15 +1,4 @@
 import { destroyEnvironment } from "../environment";
```
Comment (dpwatrous) [inline packages/bonito-core/src/__tests__/setup-tests.ts:1]: I'm all for factoring out our JSDOM stubs/mocks, but at least some of that needs to be in bonito-core since HttpClient needs both fetch() and Blob.

File: packages/bonito-ui/src/environment/mock-browser-environment.ts [hunk 13]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable @typescript-eslint/no-empty-function */
```
Comment (dpwatrous) [inline packages/bonito-ui/src/environment/mock-browser-environment.ts:1]: Why is this needed?

File: util/mock-style.js [hunk 14]
```diff
@@ -0,0 +1 @@
+module.exports = {};
```
Comment (dpwatrous) [inline util/mock-style.js:1]: Can we move this into common-config? `util` wasn't meant to contain source code, but rather be a parent folder for utility packages.

File: packages/react/src/vm-extension/mock-vm-ext-items.ts [hunk 15]
```diff
@@ -0,0 +1,92 @@
+import { VmExtItem } from "./vm-extension-list";
+
+export const succeededExtItem: VmExtItem = {
```
Comment (dpwatrous) [inline packages/react/src/vm-extension/mock-vm-ext-items.ts:3]: Can we move this file to the __tests__ folder if this isn't referenced outside of the local tests?

**Review by dpwatrous — CHANGES_REQUESTED** [review]

File: packages/bonito-core/src/__tests__/setup-tests.ts (hunk 12 above)
Comment (wanghoppe) [reply packages/bonito-core/src/__tests__/setup-tests.ts:1]: Yeah, that's the problem, do you think it's OK to move `jsdom-mocks.ts` into bonito-core or just the blob and fetch?

File: packages/bonito-ui/src/components/data-grid/data-grid.tsx [hunk 16]
```diff
@@ -102,7 +102,14 @@ export interface DataGridColumn {
     /**
      * Custom renderer for cell content, instead of the default text rendering.
      */
-    onRender?: (item?: unknown, index?: number, column?: IColumn) => unknown;
+    onRender?: (
+        // can't use unknown here because it's a function's parameter, typescript
```
[the remark this answers is not in the document]
Comment (wanghoppe) [reply packages/bonito-ui/src/components/data-grid/data-grid.tsx:106]: will remove

File: packages/bonito-ui/src/environment/mock-browser-environment.ts (hunk 13 above)
Comment (wanghoppe) [reply packages/bonito-ui/src/environment/mock-browser-environment.ts:1]: will remove

File: packages/react/src/vm-extension/mock-vm-ext-items.ts (hunk 15 above)
Comment (wanghoppe) [reply packages/react/src/vm-extension/mock-vm-ext-items.ts:3]: This is used in playground as well to display VM extension list demo page, so distributing it as well.

File: packages/bonito-core/src/__tests__/setup-tests.ts (hunk 12 above)
Comment (dpwatrous) [reply packages/bonito-core/src/__tests__/setup-tests.ts:1]: I'm a little torn, but if we want to keep saying that bonito-core can work in Node.js we should probably split them. Node does have Blob and fetch, but won't likely ever implement `matchMedia` or `HTMLCanvasElement`.

File: packages/react/src/vm-extension/mock-vm-ext-items.ts (hunk 15 above)
Comment (dpwatrous) [reply packages/react/src/vm-extension/mock-vm-ext-items.ts:3]: Ah, I missed that. Can we move this to a folder which makes it clear this is for tests? Could be either the service package's test-util folder, or one in the react package. Just want to be careful mixing test data and code with production code (even though I know there's an overlap when it comes to the playground).

Comment (wanghoppe) [reply packages/react/src/vm-extension/mock-vm-ext-items.ts]: moved to @batch/ui-react package's test-util

File: packages/react/src/vm-extension/__tests__/vm-extension-list.spec.tsx [hunk 17]
```diff
@@ -0,0 +1,90 @@
+// FILEPATH: /c:/Users/hoppe/work/sandbox/BatchExplorer/packages/react/src/vm-extension/__tests__/vm-extension-list.spec.tsx
```
Comment (dpwatrous) [inline packages/react/src/vm-extension/__tests__/vm-extension-list.spec.tsx:1]: Remove this line

File: packages/react/src/test-util/mock-vm-ext-items.ts [hunk 18]
```diff
@@ -0,0 +1,92 @@
+import { VmExtItem } from "../vm-extension/vm-extension-list";
+
+export const succeededExtItem: VmExtItem = {
```
Comment (dpwatrous) [inline packages/react/src/test-util/mock-vm-ext-items.ts:3]: Seems like we should have some extension data in BatchFakeSet

File: packages/react/src/vm-extension/__tests__/vm-extension-details-panel.spec.tsx [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+        const { getByText } = render(
+            <VmExtensionDetailsPanel
+                isOpen={false}
+                vme={succeededExtItem}
+                onDismiss={onDismiss}
+            />
+        );
+        expect(() =>
+            getByText(translate("lib.react.vmExtension.extensionProperties"))
+        ).toThrow();
```
Comment (dpwatrous) [inline packages/react/src/vm-extension/__tests__/vm-extension-details-panel.spec.tsx:36]: Can we assert positively that there's an empty or hidden element or something like that instead of just checking that the text doesn't exist?

Comment (wanghoppe) [reply packages/react/src/vm-extension/__tests__/vm-extension-details-panel.spec.tsx:36]: This one is hard. When the panel opens, it actually adds a <div> under <body> and removes it when closes.

Comment (dpwatrous) [reply packages/react/src/vm-extension/__tests__/vm-extension-details-panel.spec.tsx:36]: Happy to follow up on this, but this isn't a total blocker. I'll just resolve and approve since this PR has been out for so long.

**Review by dpwatrous — APPROVED** [review]

File: util/common-config/jest-common.js [hunk 20]
```diff
@@ -55,6 +59,14 @@ module.exports = {
                 ],
             ],
             globals: {
+                "ts-jest": {
+                    // js-with-ts preset require "allowJs": true in tsconfig.json
+                    // so making a "tsconfig.test.json" to override it
+                    tsconfig: "config/tsconfig.test.json",
```
Comment (dpwatrous) [inline util/common-config/jest-common.js:65]: Can we either change the `createConfig()` function to not take in a tsconfig, or pass `config/tsconfig.test.json` in for each package?

**Review by dpwatrous — APPROVED** [review]

File: util/common-config/jest-common.js (hunk 20 above)
Comment (wanghoppe) [reply util/common-config/jest-common.js:65]: We are passing a `tsconfig` object read from `tsconfig.json` for each package, and the only usage of this object is to pass `compilerOptions.paths` into jest's `moduleMapper`. But I didn't find any `tsconfig.json` we read from has the `compilerOptions.paths` field, do you think it's OK to remove the `tsconfig` object argument for `createConfig()`?

Comment (dpwatrous) [reply util/common-config/jest-common.js:65]: It looks like `ts-jest`'s tsconfig property supports an inline object, so I guess I'd lean toward using the passed-in object. Then we'd just have to update each jest.config.js from `require("./tsconfig.json")` to `require("./config/tsconfig.test.json")`. That also keeps the `config/ts*.json` directory structure a convention rather than something jest-common.js needs to work.

Comment (wanghoppe) [reply util/common-config/jest-common.js:66]: Removed `tsconfig` parameter for `createConfig()`

**Review by dpwatrous — DISMISSED** [review]

**Review by dpwatrous — APPROVED** [review]
